pkg/util/annotationUtil.go

```diff
@@ -0,0 +1,151 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"sync"
+)
+
+//Client request grafana API on a set of resource paths.
+type client struct {
+	// base is the root URL for all invocations of the client
+	baseUrl url.URL
+	client *http.Client
+}
+
+//Annotation is a specification of the desired behavior of adding annotation
+type Annotation struct {
+	dashboardId int
+	panelId int
+	timestampInMilliSec int64
+	tags []string
+	text string
+}
+
+//NewClient creats a new grafanaClient. This client performs rest functions
+//such as Get, Post on specified paths.
+func NewClient(grafanaUrl string, userName string, password string, prometheusExporterPort int) (*client, error){
+	u, err := url.Parse(grafanaUrl)
+	if err != nil {
+		return nil, err
+	}
+
+	initFunc(prometheusExporterPort)
+	u.User = url.UserPassword(userName, password)
+	return &client{
+		baseUrl: *u,
+		client: &http.Client{},
+	}, nil
+}
+
+func (annotation *Annotation) getBody() ([]byte, error) {
+	m := map[string]interface{} {
+		"dashboardId": annotation.dashboardId,
+		"panelId": annotation.panelId,
+		"time": annotation.timestampInMilliSec,
+		"isRegion": false,
+		"tags": annotation.tags,
+		"text": annotation.text,
+		"timeEnd": 0,
+	}
+
+	body, error := json.Marshal(m)
+	if error != nil {
+		return nil, error
+	}
+
+	return body, nil
+}
+
+var(
+	initedOnce sync.Once
+	counterMetric prometheus.Counter
+	annotationSubPath = "api/annotations"
+)
+
+//initFunc is called with sync.Once, we use sync.Once to keep the thread safe.
+func initFunc(port int) {
+	initedOnce.Do(func() {
+		counterMetric = initErrorMetric()
+		prometheus.MustRegister(counterMetric)
+		mux := http.NewServeMux()
+
+		l, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
+		if err != nil {
+			fmt.Fprint(os.Stderr, "listening port 8083 failed", err)
+			panic(err)
+		}
+
+		mux.Handle("/metrics", promhttp.Handler())
+		srv := &http.Server{Handler: mux}
+		go srv.Serve(l)
+	})
+}
+
+func initErrorMetric() prometheus.Counter{
+	return prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "errorcount",
+		Help: "record error count",
+		ConstLabels: map[string]string{"fortest": "true"},
+	})
+}
+
+//IncreErrorCountWithAnno increments the errorcount by 1,
+//and add the annotation to grafanan.
+func (cli *client) IncreErrorCountWithAnno(annotation Annotation) error{
```

Comment:
we don't need this method. only `addAnnotation` is needed.

Reply:
Split it as two functions
1.AddAnnotation
2.IncrErrorCount